REGRESSION (271770@main): visionOS: Significant regression on MotionMark

hortont424 opened the pull request:
#### 03a60be8929d23706595fd8fcaa834e3839912ae
<pre>
REGRESSION (271770@main): visionOS: Significant regression on MotionMark
<a href="https://bugs.webkit.org/show_bug.cgi?id=269010">https://bugs.webkit.org/show_bug.cgi?id=269010</a>
<a href="https://rdar.apple.com/122127093">rdar://122127093</a>

Reviewed by Simon Fraser.

Before 271770@main, the dynamic content scaling image buffer was a sidecar, outside
of the normal flow of the front/back buffer swapping. After that commit, it was moved
inside the main image buffer, and thus is swapped like the base scale rendering.

This had one unintended consequence: when we&apos;re discarding the back buffer
(e.g. because it contains an IOSurface that is still in-use and can&apos;t be recycled),
we will now *also* discard the dynamic content scaling image buffer, and, critically,
its resource cache.

The whole point of the resource cache is to maintain state between frames, so this is
wildly counterproductive. Once you get into a state where you are discarding
back buffers (which comes up frequently under load, like MotionMark), you&apos;re also
losing all resource caching.

Instead of trying to rearchitect the discarding code to avoid dropping the display
list image buffer, just maintain the dynamic content scaling resource cache
in a sidecar on the RemoteLayerBackingStore, like it was before, and pass it
down to the image buffer.

A subsequent patch will make a similar change for the remote rendering case.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/re/DynamicContentScalingResourceCache.h: Added.
(WebCore::DynamicContentScalingResourceCache::DynamicContentScalingResourceCache):
* Source/WebCore/platform/graphics/re/DynamicContentScalingResourceCache.mm: Added.
(WebCore::DynamicContentScalingResourceCache::create):
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.mm:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::ensureDynamicContentScalingResourceCache):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer const): Deleted.

* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
Note that it is not safe in a unified-sources world to forward-declare WebCore::IOSurface
because of the name conflict with ::IOSurface; we need to strictly order the declarations,
which is easiest to achieve by just importing our header.

Canonical link: <a href="https://commits.webkit.org/274319@main">https://commits.webkit.org/274319@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa850181fbb4a9b7258686e250d1e58537eae58b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38698 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17629 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41043 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41229 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41005 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20439 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14975 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32493 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39271 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14856 "Found 1 new test failure: fast/events/ios/select-all-with-existing-selection.html (failure)") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12901 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/34503 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42505 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35153 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38710 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/13508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11180 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36916 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/15120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5046 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
```diff
@@ -87,9 +87,13 @@ static CFDictionaryRef makeContextOptions(const DynamicContentScalingImageBuffer
+    // FIXME: We should make callers always specify a cache and have an assertion here instead
+    // of making a temporary one. RemoteLayerWithRemoteRenderingBackingStore currently does not.
+    if (!m_resourceCache)
+        m_resourceCache = bridge_id_cast(adoptCF(RECGCommandsCacheCreate(nullptr)));
```
Could `m_resourceCache` here be a `DynamicContentScalingResourceCache`? Would be nice not to see `RECGCommandsCacheCreate` more than once in the codebase.

hortont424 replied:
It could! I'll leave a note about that in the bug that I filed about the FIXME above.

hortont424 committed:
REGRESSION (271770@main): visionOS: Significant regression on MotionMark
https://bugs.webkit.org/show_bug.cgi?id=269010
rdar://122127093

Reviewed by Simon Fraser.

Before 271770@main, the dynamic content scaling image buffer was a sidecar, outside
of the normal flow of the front/back buffer swapping. After that commit, it was moved
inside the main image buffer, and thus is swapped like the base scale rendering.

This had one unintended consequence: when we're discarding the back buffer
(e.g. because it contains an IOSurface that is still in-use and can't be recycled),
we will now *also* discard the dynamic content scaling image buffer, and, critically,
its resource cache.

The whole point of the resource cache is to maintain state between frames, so this is
wildly counterproductive. Once you get into a state where you are discarding
back buffers (which comes up frequently under load, like MotionMark), you're also
losing all resource caching.

Instead of trying to rearchitect the discarding code to avoid dropping the display
list image buffer, just maintain the dynamic content scaling resource cache
in a sidecar on the RemoteLayerBackingStore, like it was before, and pass it
down to the image buffer.

A subsequent patch will make a similar change for the remote rendering case.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/re/DynamicContentScalingResourceCache.h: Added.
(WebCore::DynamicContentScalingResourceCache::DynamicContentScalingResourceCache):
* Source/WebCore/platform/graphics/re/DynamicContentScalingResourceCache.mm: Added.
(WebCore::DynamicContentScalingResourceCache::create):
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingBifurcatedImageBuffer.mm:
* Source/WebKit/Shared/RemoteLayerTree/DynamicContentScalingImageBufferBackend.mm:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::ensureDynamicContentScalingResourceCache):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer):
(WebKit::RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer const): Deleted.

* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
Note that it is not safe in a unified-sources world to forward-declare WebCore::IOSurface
because of the name conflict with ::IOSurface; we need to strictly order the declarations,
which is easiest to achieve by just importing our header.

Canonical link: https://commits.webkit.org/274319@main

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm:
```diff
@@ -254,15 +261,17 @@ void flushAndCollectHandles(HashMap<RemoteImageBufferSetIdentifier, std::unique_
 #if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
-    if (m_parameters.includeDisplayList == IncludeDisplayList::Yes)
+    if (m_parameters.includeDisplayList == IncludeDisplayList::Yes) {
+        creationContext.dynamicContentScalingResourceCache = ensureDynamicContentScalingResourceCache();
```
Maybe calling `ensureDynamicContentScalingResourceCache()` can be moved to `RemoteLayerWithInProcessRenderingBackingStore::ensureBackingStore()`. So this one can just assign 
```
creationContext.dynamicContentScalingResourceCache = m_dynamicContentScalingResourceCache;
```
And the function can stay const.
